exercises/chapter4/test/Main.purs

```diff
@@ -168,6 +164,21 @@ Note to reader: Delete this line to expand comment block -}
           $ Assert.equal (Nothing)
           $ map filename
           $ whereIs root "cat"
+      suite "Exercise - largestSmallest for root" do
+        test "has length of 2" do
+          Assert.equal 2
+            $ length 
+            $ largestSmallest root
+        test "includes /home/user/code/js/test.js" do
+          Assert.equal (Just "/home/user/code/js/test.js")
+            $ map filename 
+            $ find (\p -> filename p == "/home/user/code/js/test.js")
+            $ largestSmallest root
+        test "includes /etc/hosts" do
+          Assert.equal (Just "/etc/hosts")
+            $ map filename
+            $ find (\p -> filename p == "/etc/hosts")
+            $ largestSmallest root
```

Comment:
Let's add tests for the corner cases of one-file and no-files. The no-files case will need an empty directory added to the dummy data here:
https://github.com/purescript-contrib/purescript-book/blob/a850ae585fe49d95ade5c1184d67edfdc32f9b04/exercises/chapter4/src/Data/Path.purs#L21-L25

Also, you can simplify each cluster of "has length / includes / includes" into a single test by sorting the expected and actual results. That strategy is used for these `cartesianProduct` and `triples` tests.
https://github.com/purescript-contrib/purescript-book/blob/a850ae585fe49d95ade5c1184d67edfdc32f9b04/exercises/chapter4/test/Main.purs#L81-L111
The `cartesianProduct` tests use an additional `testcp` convenience helper-function. Something like that might be useful here too, and it should be simpler to write since it only needs to handle `Array Path` rather than a more generic `Array a`. But if any of this testing code involves too much unknown territory, feel free to circle back later after you tackle more chapters in the book. Alternatively, I can help fill in any missing details now by answering questions in this PR or directly editing the file. I'm also available for a peer programming session if hint-driven-development seems faster. Ping me on Discourse/Slack/Zulip if you'd like to set-up a time for that.

Reply:
Hey thanks for the feedback. I think this all sounds pretty doable. My only question is regarding the new dummy data. Is the idea to add a directory with no files to the existing `root` `Path`? If so, do you have a recommendation on how to acquire the directories with one and no files from `root` to provide to `largestSmallest`? I don't think there is currently a function for this chapter that finds a `Path` by `filename`. I could write one but wanted to verify that this is the right direction.

Reply:
That's a good question. My original `root` proposal is not so straightforward.

I now think the easiest option is to create these corner-case `Path` trees locally in `test/Main.purs`.

To grab these paths from within `root`, we'd need a `whereIs` that returns the found path populated with children, and re-writing that in the test framework is a bit chicken&egg.

Reply:
I just pushed up a couple of commits using the approach outlined above. Happy to make any additional changes.